nmstatectl/nmstatectl.py

```diff
@@ -98,6 +125,77 @@ def _filter_state(state, whitelist):
     return state
 
 
+def edit(args):
+    state = _filter_state(netinfo.show(), args.only)
+
+    if args.yaml:
+        suffix = '.yaml'
+        txtstate = pretty_yaml(state)
+    else:
+        suffix = '.json'
+        txtstate = pretty_json(state)
+    with tempfile.NamedTemporaryFile(suffix=suffix,
+                                     prefix='nmstate-') as statefile:
+        statefile.write(txtstate)
+        statefile.flush()
+
+        while True:
+            txtstate = run_editor(txtstate, suffix)
+            error = ""
+
+            if txtstate is None:
+                return 1
+
+            try:
+                if args.yaml:
+                    new_state = yaml.load(txtstate)
+                else:
+                    new_state = json.loads(txtstate)
+            except ValueError as e:
+                error = "Invalid JSON syntax: %s\n" % e
+            except yaml.parser.ParserError as e:
+                error = "Invalid YAML syntax: %s\n" % e
+
+            if not error and (not new_state or 'interfaces' not in new_state):
+                error = "Invalid state: should contain 'interfaces' entry.\n"
+
+            if error:
+                sys.stderr.write("ERROR: ")
+                sys.stderr.write(error)
+                response = ""
+                while response not in ("y", "n"):
+                    response = input("Try again? [y,n]:\n"
+                                     "y - yes, start editor again\n"
+                                     "n - no, throw away my changes\n"
+                                     "> ")
+                if response == "n":
+                    return 1
+            else:
+                break
+
+        print('Applying the following state: ')
+        print_state(new_state, use_yaml=args.yaml)
+
+        netapplier.apply(new_state, verify_change=args.verify)
+
+
+def run_editor(txtstate, suffix):
+    editor = os.environ.get('EDITOR', 'vi')
+    with tempfile.NamedTemporaryFile(suffix=suffix,
+                                     prefix='nmstate-') as statefile:
+        statefile.write(txtstate)
+        statefile.flush()
+
+        try:
+            subprocess.check_call([editor, statefile.name])
+            statefile.seek(0)
+            return statefile.read()
+
+        except subprocess.CalledProcessError:
+            sys.stderr.write('Error running editor, aborting...\n')
```

Comment:
Don't you want to report the exception info?

Reply:
The idea is to allow users to quit vim with `:cq` without changing the state. The exception would no be interesting then. Not sure if there are other error cases where it would be.